### PR TITLE
Add padding_free packed forward for GRPO and RLOO per-token log-probs

### PR DESCRIPTION
**Pull request by behroozazarkhalili**
GRPO and RLOO log-probability forwards spent compute on padded tokens. I added optional padding-free forwards that pack real tokens, reset positions per sample, and restore the padded output layout, with checks for unsupported configurations. CPU tests cover output values, gradients, and guards, and GPU tests compare packed and dense outputs and train with finite loss. Prefix sharing remains open on #3549.

---

> [!NOTE]
> **Medium Risk**
> Changes the core log-prob forward path when enabled; incorrect packing or models that ignore position-id resets could silently mix samples or skew policy gradients, though defaults stay off and guards limit unsupported setups.
> 
> **Overview**
> Adds optional **`padding_free=True`** to **GRPO** and **RLOO** so per-token log-probability forwards pack each micro-batch’s non-padding tokens into one sequence with per-sample **position-id resets** (no attention mask), then **scatter** log-probs and entropies back into the usual padded `(batch, completion_len)` layout so losses and metrics stay the same.
> 
> **`GRPOConfig` / `RLOOConfig`** expose the flag; trainers validate **Flash Attention** on policy and reference models, reject **VLMs**, and GRPO also rejects **`use_liger_kernel`**. Docs describe usage, requirements, and models where position resets do not isolate samples.
> 
> New tests check packed vs dense agreement, backward grads, validation errors, and a slow GPU training path (xfail where flash kernels are broken).
> 
> <sup>Reviewed by [Cursor Bugbot](https://cursor.com/bugbot) for commit 9657bbb801db2acb0400fda919bbd9f1ab13084e. Bugbot is set up for automated code reviews on this repo. Configure [here](https://www.cursor.com/dashboard/bugbot).</sup>

### CODE REVIEW

File: trl/trainer/grpo_trainer.py [hunk 1]
```diff
@@ -1527,15 +1573,26 @@ def _get_per_token_logps_and_entropies(
...
             outputs = model(**model_inputs)
             logits = outputs.logits
-            # Exclude the last value: it corresponds to the next token pred
-            logits = logits[:, :-1, :]  # (B, L-1, H)
-            # Only keep the last logits_to_keep. For model that support logits_to_keep, this is a no-op.
-            logits = logits[:, -logits_to_keep:, :]  # (B, logits_to_keep, H)
+            if self.padding_free:
+                if "logits_to_keep" not in self.model_kwarg_keys:
+                    logits = logits[:, keep_idx - 1, :]  # (1, K, H)
+                completion_ids = model_inputs["input_ids"][:, keep_idx]
```
[the remark this answers is not in the document]
Comment (behroozazarkhalili) [reply trl/trainer/grpo_trainer.py:1584]: Confirmed by execution: a row whose prompt is all padding gave a different first-token log-prob on the packed path (-11.875 vs -11.9375 dense), because the predecessor index wraps for the first sample and points into the previous sample otherwise. adfff290 makes the packed forward raise for a sample with no prompt token (its first completion token has no in-row predecessor; the dense path scores it from a padding position, which is not a value worth reproducing), with a test in both trainers.

**Comment by behroozazarkhalili** [comment]
Re-ran the GPU oracle on the current head (9b080d3e) after the merge-forwards, one H100, `kernels-community/flash-attn2`, torch 2.9, SLURM job 58250726. Two samples of different lengths packed into one row, every real completion token compared with the dense forward:

| trainer | max abs diff, log-probs | max abs diff, entropies | chunk of 1 vs full | real tokens |
|---|---|---|---|---|
| GRPOTrainer | 0.0 | 0.0 | 0.0 | 38 |
| RLOOTrainer | 0.0 | 0.0 | 0.0 | 38 |

Negative control with the position resets removed: max abs diff 6.25e-2 on the same inputs, so the oracle does fail when isolation is broken. Two training steps each for GRPO and RLOO with `padding_free=True` finished with finite losses and all 25 parameters updated. Nothing in the body changes; this is only a confirmation that the numbers hold on the head that will merge.

File: trl/trainer/grpo_trainer.py [hunk 2]
```diff
@@ -823,6 +834,23 @@ def get_reward(environments, _env_type=env_type, **kwargs):
...
+        self.padding_free = args.padding_free
+        if self.padding_free:
+            attn_implementation = model.config._attn_implementation.split("@")[0].split(":")[0]
+            if attn_implementation not in FLASH_ATTENTION_VARIANTS:
+                raise ValueError(
+                    f"`padding_free=True` requires a Flash Attention implementation, got {attn_implementation!r}. The "
+                    "packed forward carries no attention mask and separates the samples through position-id resets, "
+                    f"which only these implementations honor: {sorted(FLASH_ATTENTION_VARIANTS)}. Set "
+                    "`attn_implementation` in the model configuration or disable `padding_free`."
+                )
```
[the remark this answers is not in the document]
Comment (behroozazarkhalili) [reply trl/trainer/grpo_trainer.py:846]: Confirmed, and fixed in 00e2028d.

The gap is real but narrower than the description suggests. When `model` is a string, `args.model_init_kwargs` reaches both loads, so the reference model inherits `attn_implementation` and the policy guard covers it. When the caller passes an already-instantiated model, `model_init_kwargs` is ignored with a warning (`grpo_trainer.py:357`), so the reference model is created from the path with nothing, and I measured that default:

```
create_model_from_path('trl-internal-testing/tiny-Qwen3ForCausalLM')            -> sdpa
create_model_from_path(..., attn_implementation='eager')                        -> eager
```

That reference model then goes through `_get_per_token_logps_and_entropies` at `grpo_trainer.py:2802`, the same packed forward, so the KL term would read contaminated reference log-probs.

The fix validates the reference model right after it is created, with a block byte-identical in `grpo_trainer.py` and `rloo_trainer.py`. Each trainer gains `test_padding_free_requires_flash_attention_for_reference_model`, which passes a caller-instantiated model with Flash Attention on the policy and `beta=0.1` so a reference model is actually built. Both pass; with the guard deleted, only the new test fails.

File: trl/trainer/grpo_trainer.py [hunk 3]
```diff
@@ -1556,14 +1618,26 @@ def _full_logits_logps(
...
             # Divide logits by sampling temperature.
             # See https://huggingface.co/blog/the_n_implementation_details_of_rlhf_with_ppo#policy-training-implementation-details
             logits = logits / self.temperature
             logps = selective_log_softmax(logits, completion_ids)  # compute logprobs
+            if self.padding_free:
+                # Scatter the packed values back into the padded (b, logits_to_keep) layout the callers expect.
+                target_mask = completion_mask_batch[:, -logits_to_keep:]
+                packed_logps = logps
+                logps = torch.zeros(target_mask.shape, dtype=packed_logps.dtype, device=packed_logps.device)
+                logps[target_mask] = packed_logps[0]
```
[the remark this answers is not in the document]
Comment (behroozazarkhalili) [reply trl/trainer/grpo_trainer.py:1640]: Checked this by running the packed path on a chunk whose completion tokens are all masked (the `mask_truncated_completions` case), on `trl-internal-testing/tiny-Qwen3ForCausalLM` with chunk sizes 1 and 2, with and without `logits_to_keep`, logps and entropies. `backward()` succeeds in every case and the gradient is zero, which matches the dense path.

The scatter is an in-place `index_put_`, and autograd records it even when the mask selects no element, so the zeros tensor leaves the block with `IndexPutBackward0` and `requires_grad=True` rather than detached. The same block is byte-identical in `rloo_trainer.py`, so the same holds there. No change made.
